File: SWGTK/tests/text_cpp/TextTest.cpp
#include <swgtk/Utility.hpp>
#include <TextTest.hpp>
#include <swgtk/App.hpp>

namespace swgtk {
	bool TextTest::Create(Scene& scene) {
		constexpr auto colorDefault = 255u;

		_app = scene.AppRoot();
		_render = scene.AppRenderer<Simple2DRenderer>();

		_mouse.texture = _render->LoadPlainWrapText("Hello\nWorld!", 0, SDL_Color{colorDefault, 0u, 0u, colorDefault});

		FontGroup::SetFontStyle(_app->GetDefaultFont(), FontStyle::Underlined);
		
		_background = _render->LoadLCDWrapText("EAT!\nSLEEP!\nCODE!");

		return true;
	}

	bool TextTest::Update([[maybe_unused]] Scene& s, const float dt) {
		// NOLINTBEGIN(cppcoreguidelines-avoid-magic-numbers) - Reason: It's pointless to create constants for this test.

		_mouse.pos = _app->GetMousePos();
		_mouse.angle += dt * 2.0;

		if(_mouse.angle > math::pi2) { _mouse.angle -= (math::pi2); }

		auto [width, height] = _mouse.texture.GetSize();

		auto rect = SDL_FRect{};

		rect.w = width * 2.0f;
		rect.h = height * 2.0f;

		rect.x = _mouse.pos.x - (rect.w / 2.0f);
		rect.y = _mouse.pos.y - (rect.h / 2.0f);

		_render->BufferClear();

		_render->DrawTexture(_background);

		// Rotating in SDL3 is in degrees...
		_render->DrawTexture(_mouse.texture, std::nullopt, rect, math::radiansToDegrees(_mouse.angle));


		// NOLINTEND(cppcoreguidelines-avoid-magic-numbers)

		return true;
	}

}


int main([[maybe_unused]]int argc, [[maybe_unused]]const char** argv) {
	constexpr auto w = 800;
	constexpr auto h = 600;

	if(swgtk::App app; app.InitGraphics("Text Test", w, h, swgtk::Simple2DRenderer::Create())) {
		app.RunGame<swgtk::TextTest>();
	}
}

File: SWGTK/tests/test_lua/TextTest.lua
local mouse = {}
local background = {}

function OnCreate()
    mouse.texture = swgtk.Render:LoadBlendedWrapText("Hello\nLua!", 0, swgtk.Colori.new(255, 0, 0))
    mouse.angle = 0.0

    swgtk.App:SetFullscreen(true)

    swgtk.Fonts.SetFontStyle(swgtk.Fonts:GetDefaultFont(), swgtk.FontStyle.Italic)

    background = swgtk.Render:LoadShadedWrapText("Lua\nRocks!!", 0, swgtk.Colori.new(0, 0, 0), swgtk.Colori.new(0, 255, 0))

    return true
end

function OnUpdate(dt)
    mouse.pos = swgtk.App:GetMousePos()
    mouse.angle = mouse.angle + (swgtk.math.pi * dt)

    if mouse.angle > swgtk.math.pi2 then
        mouse.angle = mouse.angle - swgtk.math.pi2
    end

    if swgtk.App:IsKeyReleased(swgtk.KeyCode.Space) then
        swgtk.App:SetFullscreen(false)
    end

    local width, height = mouse.texture:GetSize()
    local rect = swgtk.Rectf.new()

    rect.w = width * 2.0
    rect.h = height * 2.0

    rect.x = mouse.pos.x - (rect.w / 2.0)
    rect.y = mouse.pos.y - (rect.h / 2.0)

    swgtk.Render:BufferClear(swgtk.Colorf.new(0.0, 0.0, 0.0))

    swgtk.Render:DrawTexture(background)
    swgtk.Render:DrawTextureRotated(mouse.texture, nil, rect, swgtk.math.radiansToDegrees(mouse.angle))

    return true
end